docs/proposals/improve-status-in-CAPI-resources.md

```diff
@@ -0,0 +1,958 @@
+| `AvailableReplicas` (deprecated)     | `BackCompatibilty.AvailableReplicas` (renamed) (deprecated) | (removed)                           |
+| `FailureReason` (deprecated)         | `BackCompatibilty.FailureReason` (renamed) (deprecated)     | (removed)                           |
+| `FailureMessage` (deprecated)        | `BackCompatibilty.FailureMessage` (renamed) (deprecated)    | (removed)                           |
+| `Conditions`                         | `BackCompatibilty.Conditions` (renamed) (deprecated)        | (removed)                           |
+| `ExperimentalConditions` (new)       | `Conditions` (renamed)                                      | `Conditions`                        |
+| `UpToDateReplicas` (new)             | `UpToDateReplicas`                                          | `UpToDateReplicas`                  |
+| other fields...                      | other fields...                                             | other fields...                     |
+
+Notes:
+- The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).
+  Fields in this struct are used for supporting down conversions, thus providing users relying on v1beta1 APIs additional buffer time to pick up the new changes.
+- This proposal is using `UpToDateReplicas` instead of `UpdatedReplicas`; This is a deliberated choice to avoid 
+  confusion between update (any change) and upgrade (change of the Kubernetes versions).
+- Also `AvailableReplicas` will determine Machine's availability by reading Machine.Available condition instead of
+  computing availability as of today, however in this case the semantic of the field is not changed
+
+TODO: check `FullyLabeledReplicas`, do we still need it?
+
+#### MachineSet (New)Conditions
+
+| Condition        | Note                                                                                                             |
+|------------------|------------------------------------------------------------------------------------------------------------------|
+| `ReplicaFailure` | This condition surfaces issues on creating a Machine replica in Kubernetes, if any. e.g. due to resource quotas. |
+| `MachinesReady`  | This condition surfaces detail of issues on the controlled machines, if any.                                     |
+| `ScalingUp`      | True if available replicas < desired replicas                                                                    |
+| `ScalingDown`    | True if replicas > desired replicas                                                                              |
+| `UpToDate`       | True if all the Machines controlled by this MachineSet are up to date (replicas = upToDate replicas)             |
+| `Remediating`    | True if there is at least one Machine controlled by this MachineSet that is not passing health checks            |
```

Comment:
"Remediating" is a really _active_ sounding condition: it communicates to me that a problem is being fixed.

But the description of it is a little different: a problem has been identified.

It seems like these should either be two distinct conditions to represent (a) that a machine is not passing health checks and (b) that remediation is in progress. Or align the name and description to either (a) or (b).

Reply:
In Cluster API the fact that a problem has been identified by a MachineSet also implies that remediation has been attempted.

Considering this, I think the current description is ok because it represents what is actually happening; also it relates to something the user can validate by simply looking at machines and it is similar/consistent with other conditions e.g. `ScalingDown` condition being described as true if replicas > desired replicas.

But happy to consider improvements to the descriptions, if any.

What instead I don't think we should do is to introduce another condition; this will probably introduce more confusion than benefits. 

Expanding a little bit on this last point, I don't think it matters to users if something happens one or two reconcile loops after the trigger for this action is detected, and in case the action cannot happen/is intentionally delayed, this should surface in the condition message, e.g. remediation not possible because...

Reply:
Now the description of the condition is "This condition surfaces details about ongoing remediation of the controlled machines, if any", which should address any doubt.